File: aexample/src/main/java/org/judal/examples/java/jdbc/E18_FetchByRelationalQuery.java
package org.judal.examples.java.jdbc;

import org.junit.Test;

import java.math.BigDecimal;
import java.sql.Date;
import java.sql.Types;
import java.util.Calendar;

import org.judal.storage.table.RecordSet;
import org.judal.examples.java.model.array.Course;
import org.judal.examples.java.model.map.Student;
import org.judal.storage.Param;
import org.judal.storage.java.RelationalQuery;
import org.judal.storage.query.Predicate;
import org.judal.storage.table.ColumnGroup;

import static org.judal.storage.query.Operator.LIKE;
import static org.judal.storage.query.Operator.GT;
import static org.judal.storage.query.Connective.AND;

/**
 * Fetch records using query builders
 */
public class E18_FetchByRelationalQuery {

	@SuppressWarnings({ "unused", "deprecation" })
	@Test
	public void demo() throws Exception {
		
		setUp();
		
		// SELECT id_student, first_name , last_name, date_of_birth FROM student WHERE last_name LIKE 'S%' AND date_of_birth > '1980-01-01'
		try (RelationalQuery<Student> qry = new RelationalQuery<>(Student.class)) {
			qry.and("last_name", LIKE, "S%").and("date_of_birth", GT, new Date(80, 0, 1));
			qry.setResult(new ColumnGroup("id_student","first_name","last_name","date_of_birth"));
			RecordSet<Student> s80 = qry.fetch();
			for (Student s : s80) {
				int id = s.getId();
				String firstName = s.getFirstName();
				String lastName = s.getLastName();
				Calendar dob = s.getDateOfBirth();
			}
		}

		// SELECT * FROM course WHERE dt_created > '2017-05-01' AND price > 1000
		try (RelationalQuery<Course> qry = new RelationalQuery<>(Course.class)) {
			Predicate prd = qry.newPredicate(AND, GT, new Param[] {
				new Param("dt_created", 1, new Date(117, 4, 1)),
				new Param("price", Types.DECIMAL, 2, new BigDecimal("1000")) });
			qry.setFilter(prd);
			qry.setResult("*");
			RecordSet<Course> c17 = qry.fetch();
			for (Course c : c17) {
				int id = c.getId();
				String cCode = c.getCode();
			}
		}

		tearDown();
	}

	public static void setUp() throws Exception {
		E10_WriteCSVDataIntoTheDatabase.setUp();
		E10_WriteCSVDataIntoTheDatabase.insertCoursesIntoDatabase();
		E10_WriteCSVDataIntoTheDatabase.insertStudentsIntoDatabase();
		E10_WriteCSVDataIntoTheDatabase.assignStudentsToCoursesIntoDatabase();
	}

	public static void tearDown() throws Exception {
		E10_WriteCSVDataIntoTheDatabase.tearDown();
	}
}

File: aexample/src/main/scala/org/judal/examples/scala/jdbc/E18_FetchByRelationalQuery.scala
package org.judal.examples.scala.jdbc

import org.junit.Test

import java.sql.Date
import java.util.Calendar

import org.judal.storage.table.RecordSet
import org.judal.storage.scala.RelationalQuery
import org.judal.storage.table.ColumnGroup

import org.judal.storage.query.Operator.LIKE
import org.judal.storage.query.Operator.GT

import org.judal.examples.scala.model.Student

import org.judal.Using._

/**
 * Fetch records using a query builder
 */
class E18_FetchByRelationalQuery {

	@Test
	def demo() = {
		
		E18_FetchByRelationalQuery.setUp()
		var qry : RelationalQuery[Student] = null
		using (qry) {
		  qry = new RelationalQuery[Student](classOf[Student])
			qry.and("last_name", LIKE, "S%").and("date_of_birth", GT, new Date(80, 0, 1))
			qry.setResult(new ColumnGroup("id_student","first_name","last_name","date_of_birth"))
			for (s <- qry.fetch) {
				val id = s.getId
				val firstName = s.getFirstName
				val lastName = s.getLastName
				val dob = s.getDateOfBirth
			}
		}

		E18_FetchByRelationalQuery.tearDown()
	}

}


object E18_FetchByRelationalQuery {

	def setUp() = {
		E10_WriteCSVDataIntoTheDatabase.setUp()
		E10_WriteCSVDataIntoTheDatabase.insertCoursesIntoDatabase()
		E10_WriteCSVDataIntoTheDatabase.insertStudentsIntoDatabase()
		E10_WriteCSVDataIntoTheDatabase.assignStudentsToCoursesIntoDatabase()
	}

	def tearDown() = {
		E10_WriteCSVDataIntoTheDatabase.tearDown()
	}

}